### ISSUE TITLE
Azure: The system cannot find the path specified.

### ISSUE THREAD

**Issue by 3nth** [madskristensen/WebCompiler#146]
I'm getting an error only during build while deploying to azure. Error suggests that the less file is not found, however I have confirmed it's in the expected place via the kudu backend. Running the failed command via kudu process explorer results in same error.

Local builds work just fine. 

Standard setup for WebCompiler. BuildWebCompiler 1.8.270 installed. 

What did resolve this was to include the output files (css & min.css) in the git repo, but I was expecting that with this tooling in place it should be possible to only include the less files and ignore the generated css files.

Deployment log:
remote:   WebCompiler: Begin compiling compilerconfig.json
remote:   WebCompiler installing updated versions of the compilers...
remote: D:\home\site\repository\Ebisu.Web.Portal\Content\site.less : WebCompiler error 0: The system cannot find the path specified. [D:\home\site\repository\Ebisu.Web.Portal\Ebisu.Web.Portal.csproj]
remote:   WebCompiler: Done compiling compilerconfig.json
remote: Failed exitCode=1, command="D:\Program Files (x86)\MSBuild\14.0\Bin\MSBuild.exe" "D:\home\site\repository\Ebisu.Web.Portal\Ebisu.Web.Portal.csproj" /nologo /verbosity:m /t:Build /t:pipelinePreDeployCopyAllFilesToOneFolder /p:_PackageTempDir="D:\local\Temp\8d2df2a25231918";AutoParameterizationWebConfigConnectionStrings=false;Configuration=Release;UseSharedCompilation=false /p:SolutionDir="D:\home\site\repository\.\\"
remote: An error has occurred during web site deployment.

**Comment by nozzlegear** [comment]
I'm getting the same error, deploying a web app to Azure through GitHub continuous deployment. The less file is there, but WebCompiler can't find it.

```
  WebCompiler: Begin compiling compilerconfig.json
  WebCompiler installing updated versions of the compilers...
D:\home\site\repository\PE-Shopify-App\PE-Shopify-App\src\pages\product.less : WebCompiler error 0: The system cannot find the path specified. [D:\home\site\repository\PE-Shopify-App\PE-Shopify-App\PE-Shopify-App.csproj]
  WebCompiler: Done compiling compilerconfig.json
Failed exitCode=1, command="D:\Program Files (x86)\MSBuild\14.0\Bin\MSBuild.exe" "D:\home\site\repository\PE-Shopify-App\PE-Shopify-App\PE-Shopify-App.csproj" /nologo /verbosity:m /t:Build /t:pipelinePreDeployCopyAllFilesToOneFolder /p:_PackageTempDir="D:\local\Temp\8d2e8c5a64003bf";AutoParameterizationWebConfigConnectionStrings=false;Configuration=Release;UseSharedCompilation=false /p:SolutionDir="D:\home\site\repository\PE-Shopify-App\\"
An error has occurred during web site deployment.
```

**Comment by johannsig** [comment]
Also having this problem, entire log follows:

> Command: "D:\home\site\deployments\tools\deploy.cmd"
> Handling .NET Web Application deployment.
> All packages listed in packages.config are already installed.
  
> WebCompiler: Begin compiling compilerconfig.json
> WebCompiler installing updated versions of the compilers...
> D:\home\site\repository\StarborneWeb\Content\Site.less : WebCompiler error 0: The system cannot find the path specified. [D:\home\site\repository\StarborneWeb\StarborneWeb.csproj]
  WebCompiler: Done compiling compilerconfig.json
Failed exitCode=1, command="D:\Program Files (x86)\MSBuild\14.0\Bin\MSBuild.exe" "D:\home\site\repository\StarborneWeb\StarborneWeb.csproj" /nologo /verbosity:m /t:Build /t:pipelinePreDeployCopyAllFilesToOneFolder /p:_PackageTempDir="D:\local\Temp\8d2e9c2a2f5bde2";AutoParameterizationWebConfigConnectionStrings=false;Configuration=Release;UseSharedCompilation=false /p:SolutionDir="D:\home\site\repository\.\\"
>An error has occurred during web site deployment.
>Handling .NET Web Application deployment.\r\nAll packages listed in packages.config are already installed.\r\n  \r\n  WebCompiler: Begin compiling compilerconfig.json\r\n  WebCompiler installing updated versions of the compilers...\r\nD:\home\site\repository\StarborneWeb\Content\Site.less : WebCompiler error 0: The system cannot find the path specified. [D:\home\site\repository\StarborneWeb\StarborneWeb.csproj]\r\n  WebCompiler: Done compiling compilerconfig.json\r\nFailed exitCode=1, command="D:\Program Files (x86)\MSBuild\14.0\Bin\MSBuild.exe" "D:\home\site\repository\StarborneWeb\StarborneWeb.csproj" /nologo /verbosity:m /t:Build /t:pipelinePreDeployCopyAllFilesToOneFolder /p:_PackageTempDir="D:\local\Temp\8d2e9c2a2f5bde2";AutoParameterizationWebConfigConnectionStrings=false;Configuration=Release;UseSharedCompilation=false /p:SolutionDir="D:\home\site\repository\.\\"\r\nAn error has occurred during web site deployment.\r\nC:\Program Files (x86)\SiteExtensions\Kudu\48.41029.1889\bin\scripts\starter.cmd "D:\home\site\deployments\tools\deploy.cmd"

**Comment by whitebear-gh** [comment]
I have exactly the same issue. Could anyone take a look on it?

**Comment by johannsig** [comment]
To get around the issue I publish to Azure directly from the Visual Studio Project itself, using a publish profile from the Azure project. That works nicely for my one-man project, but probably won't help anyone trying to configure automatic builds from version control.

**Comment by phillipsj** [comment]
I am having this same issue.

**Comment by phillipsj** [comment]
So I believe this issue is being generated by line 14 in CompilerService.cs.  It is trying to get the temp folder, but since Azure is locked down I don't think it can get access and I get the "The system cannot find the path specified." error.  I then dropped into the Azure Portal command prompt and tried finding what location was set for the TMP variable and got the same error.

**Comment by madskristensen** [comment]
@phillipsj Do you know of a mechanism other than temp folder that would work for both Azure and on a local dev machine? Perhaps @davidebbo knows how to use TEMP folders on Azure?

**Comment by davidebbo** [comment]
In Azure Web Apps, TEMP/TMP are set to `TEMP=D:\local\Temp`, which is a writable folder. So in theory, any logic that relies on them should just work. We'd need to dig deeper to identify exactly what's blocking.

**Comment by madskristensen** [comment]
@davidebbo Do you think that isn't reflected in the .NET API `Path.GetTempPath()` and I should use another mechanism to get the TEMP path? If so, which do you think would be best?

**Comment by davidebbo** [comment]
`Path.GetTempPath()` works fine. I just tested it in a one line Console app running in Kudu Console, and it returned `D:\local\Temp\`.

Unlike `Environment.GetFolderPath()` (e.g. to get user profile), which has complex shell logic, I think `Path.GetTempPath()` is a pretty straightforward wrapper over the env var.

**Comment by sebvst** [comment]
Did anyone figure out what causes the problem and how to get around it? I am still getting the error that the specified path of the less file can be found.

**Comment by phillipsj** [comment]
We moved back to our TeamCity it works great there.

**Comment by pierrenallet** [comment]
@madskristensen, the missing files are the node modules. They are not unzipped. However, if you go to local/temp/webcompiler.[version] and run cmd /c prepare.cmd, the node modules are installed and you can deploy.

**Comment by eolamisan** [comment]
This issue has been open for a minute. Any progress or updates?

**Comment by whitebear-gh** [comment]
+1, any updates? We switched to dotLESS because of that... and our Interactive Developer looks like sad panda since that...

**Comment by AIRyndon** [comment]
I got the same error but on an scss file. What I did before when I got this error was redeploy and delete the wwwroot folder on Kudu; that trick didn't work this time around so I tried what @pierrenallet wrote and was thankful that I was able to push to Azure again.

**Comment by thommelutten** [comment]
I've encountered the problem as well.. The solution that @pierrenallet wrote works and I'm able to push to Azure again, but only for a while. After a day or so the error is back again.. "WebCompiler error 0: The system cannot find the path specified, and it's looking for the .csproj file.
Has anyone found a solution that lasts longer?

**Comment by AIRyndon** [comment]
@thommelutten Hello Thomas, I didn't have the issue again from my end. I think I didn't do anything else if I remember correctly. I was lucky then :)

**Comment by bwyatt** [comment]
I'm currently struggling with this issue, and I think I have a decent idea of what's happening, if not why.

It looks to me like the WebCompiler package isn't being completely loaded before you get to the build step.

My deploy.bat file looks like this:

```
:: 1. Restore nuget packages
echo Restoring nuget packages
call :ExecuteCmd dotnet restore "%DEPLOYMENT_SOURCE%\ProjectName\ProjectName.sln"
IF !ERRORLEVEL! NEQ 0 goto error

:: 2. Install bower Packages
cd "%DEPLOYMENT_SOURCE%\ProjectName\ProjectName\"
echo Installing bower packages
call bower install
IF !ERRORLEVEL! NEQ 0 goto error

:: 3. Prepare web compiler package
cd D:\local\Temp\WebCompiler*
if exist prepare.cmd (
  echo Prepare web compiler
  call prepare.cmd
  IF !ERRORLEVEL! NEQ 0 goto error
) else (
  echo Not preparing web compiler - prepare.cmd not found
)
```

When I'm deploying a new app with git push, I go through the following sequence of events:
1. On the first run, the WebCompiler folder isn't present at all, so the call to `cd` fails. Subsequently, it can't run `prepare.cmd`, so the script continues, failing later when MSBuild tries to call the compiler.
1. On the next run, the WebCompiler folder is there, but isn't fully populated yet. So the `cd` works but `prepare.cmd` isn't there yet. My script skips it and fails again at the same point later.
1. Eventually, I get lucky and the WebCompiler finishes loading `prepare.cmd` and all the other dlls into `D:\Local\Temp\WebCompiler{Version}` before my script gets to that point. When this happens, `prepare.cmd` gets executed, and all subsequent builds succeed because the node packages are there.

**TL;DR** We've got a race condition when deploying to Azure. What can we do to make sure prepare.cmd is always there before we try to call it?

**Comment by davidebbo** [comment]
@bwyatt but isn't that a clear bug in the script that it tries to cd to a folder that may not exist, and then continues regardless of success?

**Comment by bwyatt** [comment]
I suppose that's true. Right now I'm trying to figure out what files/folders I need to watch for before continuing at that point. I'm thinking that if I wait for either the `node_modules` folder to exist or for `prepare.cmd` to exist, I'll be ok.

I'll come back & update later when I have a working, repeatable solution. I'm probably going to switch the deployment script to PowerShell instead of Batch, since I'm a lot more familiar with it. Should make other future changes much easier for me.

**Comment by philringsmuth** [comment]
I don't have anything to add other than I'm experiencing the same issue and can't find a way around it. I'll explore using TeamCity instead.

**Comment by madmox** [comment]
Any update on this one? We can't setup git push deployments on our Azure web apps because of this very error. Why are the zipped files unzipped in local, but not on Azure?

**Comment by mesoamerica** [comment]
I found a workaround that allowed me to continue to work, just started VS as Administrator, opened the solution, I was able to do the operation that was throwing the error, in my case a Sync with the TFS Server

**Comment by AliSafaei1990** [comment]
it seems silly, but i had same error. i just changed my vpn server and it works!
